core/primitives/base58_codec.cpp

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// BASED ON BITCOIN IMPLEMENTATION WITH THE FOLLOWING COPYRIGHT:
+//
+// Copyright (c) 2014-2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "primitives/base58_codec.hpp"
+
+/** All alphanumeric characters except for "0", "I", "O", and "l" */
+static const char *pszBase58 =
+    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+
+static const int8_t mapBase58[256] = {
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 0,  1,  2,  3,  4,  5,  6,  7,
+    8,  -1, -1, -1, -1, -1, -1, -1, 9,  10, 11, 12, 13, 14, 15, 16, -1, 17, 18,
+    19, 20, 21, -1, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, -1, -1, -1, -1,
+    -1, -1, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, -1, 44, 45, 46, 47, 48,
+    49, 50, 51, 52, 53, 54, 55, 56, 57, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
+    -1, -1, -1, -1, -1, -1, -1, -1, -1,
+};
+
+OUTCOME_CPP_DEFINE_CATEGORY(kagome::primitives, Base58Error, e) {
+  using E = kagome::primitives::Base58Error;
+  switch (e) {
+    case E::INVALID_CHARACTER:
+      return "Invalid character in a Base58 string";
+    case E::NULL_TERMINATOR_NOT_FOUND:
+      return "Encountered a non null terminated Base58 string";
+  }
+  return "Unknown error in base58 decoder";
+}
+
+namespace kagome::primitives {
+
+  outcome::result<common::Buffer> decodeBase58(std::string_view str) noexcept {
+    // Skip leading spaces.
+    while (str.length() > 0 && std::isspace(str[0])) {
+      str = str.substr(1);
+    }
+    // Skip and count leading '1's.
+    int zeroes = 0;
+    int length = 0;
+    while (str[0] == '1') {
+      zeroes++;
+    }
+    str = str.substr(zeroes);
+
+    // Allocate enough space in big-endian base256 representation.
+    int size =
+        str.length() * 733 / 1000 + 1;  // log(58) / log(256), rounded up.
+    std::vector<unsigned char> b256(size);
+
+    // Process the characters.
+    static_assert(
+        sizeof(mapBase58) / sizeof(mapBase58[0]) == 256,
```

Comment:
I'm offering more modern way to calc size:
```suggestion
        std::size(mapBase58) == 256,
```